sklearn_numba_dpex/kmeans/drivers.py

```diff
@@ -784,6 +793,202 @@ def _get_euclidean_distances(self, X, Y):
         )
         return euclidean_distances_t.T, output_dtype
 
+    def kmeans_plusplus(self, X, sample_weight, n_clusters, random_state):
+        centers_t, center_indices, output_dtype = self._kmeans_plusplus(
+            X, sample_weight, n_clusters, random_state
+        )
+        return dpt.asnumpy(centers_t).astype(output_dtype, copy=False), dpt.asnumpy(
+            center_indices
+        )
+
+    def _kmeans_plusplus(self, X, sample_weight, n_clusters, random_state):
+        if X.shape[0] < n_clusters:
+            raise ValueError(
+                f"n_samples={X.shape[0]} should be >= n_clusters={n_clusters}."
+            )
+
+        # NB: the implementation differs from sklearn implementation with regards to
+        # sample_weight, which is ignored in sklearn, but used here.
+        # Check that this is correct ?
+        (
+            X,
+            sample_weight,
+            _,
+            compute_dtype,
+            output_dtype,
+            work_group_size,
+        ) = self._check_inputs(X, sample_weight, cluster_centers=None)
+
+        n_samples, n_features = X.shape
+
+        # heuristic taken from sklearn implementation v<1.2
+        n_local_trials = 2 + int(np.log(n_clusters))
```

Comment:
I think this might be read as implicitly implying that the heuristic has been changed in 1.2.

Is this clearer?
```suggestion
        # Same retrial heuristic as scikit-learn (at least until <1.2)
        n_local_trials = 2 + int(np.log(n_clusters))
```